File: 11/11.pl
#!/usr/bin/perl

#
# Copyright (C) 2023 Joelle Maslak
# All Rights Reserved - See License
#

use JTM::Boilerplate 'script';

use lib '.';
use Table;

use List::Util qw(all);

MAIN: {
    my $t = Table->new();
    $t->read( *ARGV, sub { split // } );

    my (@empty_row) = grep { all { $_ eq '.' } $t->get_row($_) } 0 .. ( $t->row_count() - 1 );
    my (@empty_col) = grep { all { $_ eq '.' } $t->get_col($_) } 0 .. ( $t->col_count() - 1 );

    my $matcher  = sub ($v) { return ( $v eq '#' ); };
    my @galaxies = $t->get_matching_coords($matcher);

    my $sum = get_sums( $t, 2, \@empty_row, \@empty_col, \@galaxies );
    say("Part A Sum: $sum");

    $sum = get_sums( $t, 1_000_000, \@empty_row, \@empty_col, \@galaxies );
    say("Part B Sum: $sum");
}

sub get_sums ( $t, $expansion, $empty_row, $empty_col, $galaxies ) {
    my $sum = 0;
    for ( my $i = 0; $i < scalar(@$galaxies) - 1; $i++ ) {
        for ( my $j = $i + 1; $j < scalar(@$galaxies); $j++ ) {
            $sum +=
              dist( $t, $expansion, $empty_row, $empty_col, $galaxies->[$i], $galaxies->[$j] );
        }
    }
    return $sum;
}

sub dist ( $t, $expansion, $empty_row, $empty_col, $coord1, $coord2 ) {
    my (@rows) = sort { $a <=> $b } ( $coord1->row(), $coord2->row() );
    my (@cols) = sort { $a <=> $b } ( $coord1->col(), $coord2->col() );

    my (@rowexpand) = grep { $rows[0] < $_ and $rows[1] > $_ } @$empty_row;
    my (@colexpand) = grep { $cols[0] < $_ and $cols[1] > $_ } @$empty_col;

    return $rows[1] - $rows[0] + $cols[1] - $cols[0] + ( scalar(@rowexpand) * $expansion ) +
      ( scalar(@colexpand) * $expansion ) - scalar(@rowexpand) - scalar(@colexpand);
}

File: 11/11.go
package main

import (
	"fmt"
	"os"
	"sort"
)

type DistNode struct {
	distance int
	coord    Coord
}

func main() {
	var t Table[rune]
	t.Read(os.Stdin, func(s string) []rune { return []rune(s) })

	emptyRow := make(map[int]bool)
	for i, r := range t.GetRows() {
		emptyRow[i] = r.all(func(v rune) bool { return v == '.' })
	}

	emptyCol := make(map[int]bool)
	for i, c := range t.GetCols() {
		emptyCol[i] = c.all(func(v rune) bool { return v == '.' })
	}

	galaxies := t.GetMatchingCoords(func(v rune) bool { return v == '#' })

	fmt.Printf("Part A Sum: %d\n", getSums(t, 2, emptyRow, emptyCol, galaxies))
	fmt.Printf("Part B Sum: %d\n", getSums(t, 1_000_000, emptyRow, emptyCol, galaxies))
}

func getSums(t Table[rune], expansion int, emptyRow map[int]bool, emptyCol map[int]bool, galaxies []Coord) int {
	sum := 0
	for i := range galaxies {
		for j := range galaxies {
			if i >= j {
				continue
			}
			sum += dist(t, expansion, emptyRow, emptyCol, galaxies[i], galaxies[j])
		}
	}
	return sum
}

func dist(t Table[rune], expansion int, emptyRow map[int]bool, emptyCol map[int]bool, g1 Coord, g2 Coord) int {
	rows := append(make([]int, 0, 0), g1.Row(), g2.Row())
	cols := append(make([]int, 0, 0), g1.Col(), g2.Col())

	sort.Ints(rows)
	sort.Ints(cols)

	rowExpand := 0
	for i, v := range emptyRow {
		if v && rows[0] < i && rows[1] > i {
			rowExpand++
		}
	}

	colExpand := 0
	for i, v := range emptyCol {
		if v && cols[0] < i && cols[1] > i {
			colExpand++
		}
	}

	return rows[1] - rows[0] + cols[1] - cols[0] + (rowExpand * expansion) +
		(colExpand * expansion) - rowExpand - colExpand
}
